keps/sig-node/4742-node-topology-downward-api/README.md

```diff
@@ -0,0 +1,409 @@
+A new admission plugin, `PodTopologyLabels`, will be added to kube-apiserver. When enabled, it will mutate the `pods/binding` subresource,
+adding topology labels that match those of the target Node. The Binding REST implementation will be updated to copy these labels
+from the `pods/binding` subresource to the assigned Pod's labels.
+
+Using the downward API to retrieve topology information will look similar to the following:
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  generateName: downwardapi-example-
+spec:
+  containers:
+    - name: container
+      image: registry.k8s.io/busybox
+      command: ["sh", "-c"]
+      args:
+      - while true; do
+          if [[ -e /etc/podinfo/zone ]]; then
+            echo -en '\n\n'; cat /etc/podinfo/zone; fi;
+          sleep 5;
+        done;
+      volumeMounts:
+        - name: podinfo
+          mountPath: /etc/podinfo
+  volumes:
+    - name: podinfo
+      downwardAPI:
+        items:
+          - path: "zone"
+            fieldRef:
+              fieldPath: metadata.labels['topology.kubernetes.io/zone']
+```
+
+### User Stories
+
+* As an ML engineer, I want to optimize GPU-to-GPU communication during training which requires topology-awareness in my training code.
+* As a database developer, I want to leverage topology information to improve fault tolerance of sharded databases on Kubernetes.
+* As a developer of a Kubernetes CNI plugin, I want to pass topology information down to the CNI plugin to optimize data paths for container networks.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+* Scope creep. Allowing additional node information or node label info could
+create security issues. This is mitigated by limiting the node labels
+to strictly those that are standardized through KEP-1659.
+
+* Exposing sensitive data as node labels to pods. This is mitigated by ensuring
+standard topology labels are available to Pods.
+
+* Stale data. Information obtained through node labels is like information
+attained through a configmap or secret mounted to a pod, being passed on
+creation but not guaranteed to be immutable and thus should be treated as so.
+
+
+## Design Details
+
+* A built-in Kubernetes admission plugin, `PodTopologyLabels` will be introduced in kube-apiserver
+* The `PodTopologyLabels` admission plugin is responsible for mutating `pods/binding` subresource, adding topology labels matching the target Node.
+* `PodTopologyLabels` admission will not override any existing topology labels on Pods.
+* A feature gate, `PodTopologyLabelsAdmission` will be introduced in v1.33. Alpha and disabled by default.
+The `PodTopologyLabels` admission plugin can only be set when this feature gate is enabled.
+* The Binding REST implementation will be updated to copy all labels from `pods/binding` subresource into Pods.
+At this point we do not remove labels, it is assumed that any labels in Binding are safe to expose via Downward API.
```

Comment:
one question for the future. If/when we will allow un-scheduling Pods due to things like failed devices, we will need to have a way to override these values. But we will have troubles doing it if automatic labels has no difference from the original ones. Is it too bad to start with overwriting customer labels? If we believe it will break many customers, than we need to come up with the way to differentiate these labels from customers' labels.

Reply:
This is a really good question, and again exposes a fundamentally missing piece of our APIs.

options:

1) ignore what I said and blast these labels, regardless of what the pod said before.
2) make kubelet set them on the pod when it agrees to admit it (point of no return).
3) store these somewhere else (e.g. status).
4) ?

@andrewsykim @SergeyKanzhelev do you think this question derails the KEP for this release?

Reply:
We discussed offline, given we are only allowing standard k8s labels, it should be okay to overwrite them during admission